### PR TITLE
[SPARK-28199][SS] Move Trigger implementations to Triggers.scala and avoid exposing these to the end users

### PR DESCRIPTION
**Pull request by HeartSaVioR**
## What changes were proposed in this pull request?

This patch proposes moving all Trigger implementations to `Triggers.scala`, to avoid exposing these implementations to the end users and let end users only deal with `Trigger.xxx` static methods. This fits the intention of deprecation of `ProcessingTIme`, and we agree to move others without deprecation as this patch will be shipped in major version (Spark 3.0.0).

## How was this patch tested?

UTs modified to work with newly introduced class.

### CODE REVIEW

**Comment by srowen** [comment]
Why introduce a new abstraction? This is what `Trigger.ProcessingTime` is meant to be. Just move the implementation to that class, and have the deprecated impl use it, rather than the other way around?

**Comment by HeartSaVioR** [comment]
Trigger is just a user side API, and implementation has been placed on Scala side like how `ProcessingTime` is implemented. `Once` and `Continuous` followed the same way.

That said, if we would like to proceed smoother migration with respecting the pattern what Spark has been doing, we need a new class which takes the role instead.

If we think we can remove ProcessingTime, we may not need this smoother but redundant approach. Just change case class and object of ProcessingTime to `private[sql]`, and remove deprecation.

**Comment by srowen** [comment]
Because `ProcessingTime` was deprecated in 2.2.0, I actually think it's fine and good to remove it for 3.0. Does this get simpler if you can just remove the old implementation and only use the new one? That's probably what should happen here anyway.

**Comment by HeartSaVioR** [comment]
When we decide to discontinue supporting `ProcessingTime`, looks like we have two choices:

1) Just change the scope to restrict users' access. Existing queries may fail but it is what we expected.  Least change required, but this way someone may try to hack and continue using it. Not sure we are concerning that hacky approach, though.

2) Remove the old implementation and only use the new one. Pretty clear to represent we don't support it anymore since no one could try to hack, as I'd rather not think someone applies hack to new class. More change required but simpler than current diff, maybe.

Which option do you feel better?

**Comment by srowen** [comment]
I personally favor removing the old deprecated implementation. It's probably simpler, and 3.0 is the right time to do it. I don't think this is 'critical' to continue supporting as a deprecated API, compared to many other things we've removed for similar reasons.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TriggerExecutor.scala
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TriggerExecutor.scala]: Can you remove the `ProcessingTime` import?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousExecution.scala
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousExecution.scala]: Should this use `Trigger.ProcessingTime` instead of this impl class directly or do I misunderstand?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProcessingTimeTrigger.scala
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProcessingTimeTrigger.scala]: I see, if this is basically an implementation class, I wonder if it belongs in the (unfortunately named) `Triggers.scala` class, which only now has `OneTimeTrigger`, but at least is just another implementation class too? No big deal.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousExecution.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousExecution.scala]: Trigger.xxx returns `Trigger`, whereas it requires `ProcessingTimeTrigger` (was requiring `ProcessingTime`). From user-facing API we avoid dealing with actual implementations via only exposing `Trigger` but internally we require actual implementations.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProcessingTimeTrigger.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProcessingTimeTrigger.scala]: Once we decide to move this class to `Triggers.scala`, I guess `ContinuousTrigger` has to be moved too. No big deal for me too, so please let me know which feels cleaner for you.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousExecution.scala
Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousExecution.scala]: OK, so this 'processing time' implementation requires the concrete implementation of the trigger. If so I get it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProcessingTimeTrigger.scala
Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProcessingTimeTrigger.scala]: I think I'd move it, to rationalize `Triggers.scala` and avoid another file. It doesn't matter much. I see there is `ContinuousTrigger` too but I guess it belongs in the `.continuous` subpackage.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProcessingTimeTrigger.scala]: Thanks for your voice on this. Agreed on both points, I'll leave ContinuousTrigger as it is, and move ProcessingTimeTrigger.

**Comment by HeartSaVioR** [comment]
Looks like everything sorted out. Please take a look again. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: Later we could make this non-private, but it seems like the intent is to expose through `Trigger.ProcessingTime` despite the fact that `OneTimeTrigger` is not private. Looks OK to me.

**Review by srowen — APPROVED** [review]
CC @tdas or possibly @jose-torres

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TriggerExecutor.scala
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TriggerExecutor.scala]: Please rename the variable together.
- `processingTime` -> `processingTimeTrigger`.
- `private val intervalMs = processingTime.intervalMs` -> `private val intervalMs = processingTimeTrigger.intervalMs`

**Comment by jose-torres** [comment]
Yeah, I'm also in favor of removing the deprecated implementation.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
I agree with the purpose of introducing a new case class name.
However, let's be clear. Technically, this PR is doing a kind of renaming and changing the visibility of the existing `ProcessingTime`. It would be great if we can mention that change explicitly in the PR title.

**Comment by HeartSaVioR** [comment]
Let me explain my thought on the patch.

Actually I assume the reason we deprecated `ProcessingTime` is that we don't want to expose implementations to user side. That is the only valid reason I can understand why they were deprecated and end users were encouraged to use static methods of Trigger instead. Not 100% sure. If that's the intention, we should hide other trigger implementations as well, so...

I dug the history when it's marked as deprecated, and realize there's no explanation about the reason of deprecation. No description, no review comment on deprecating these methods. So still not sure.

So based on my assumption (and my intention of the patch), it is not just renaming and changing the visibility. It finally ends up looking like that, but in fact it "removes" the deprecated class, and create the other class with applying the intention of deprecation. I will update the title of PR to include `introduce new class to replace ProcessTime`.

If my assumption is incorrect, we may need to ask to ourselves why we deprecated these methods. Honestly I can't find any other reasons of doing that.

**Comment by HeartSaVioR** [comment]
FYI, the PR which marked these methods as deprecated is here: https://github.com/apache/spark/pull/17219

**Comment by HeartSaVioR** [comment]
cc. to @tcondie hopefully who might be able to explain the reason.
Also cc. @marmbrus @tdas who reviewed that patch.

We may need to just guess if we couldn't hear any information.

**Comment by dongjoon-hyun** [comment]
Thank you for update. 

BTW, is the failure due to a flaky test case?
```
[info] - query without test harness *** FAILED *** (2 seconds, 931 milliseconds)
[info]   scala.Predef.Set.apply[Int](0, 1, 2, 3).map[org.apache.spark.sql.Row, scala.collection.immutable.Set[org.apache.spark.sql.Row]](((x$3: Int) => org.apache.spark.sql.Row.apply(x$3)))(immutable.this.Set.canBuildFrom[org.apache.spark.sql.Row]).subsetOf(scala.Predef.refArrayOps[org.apache.spark.sql.Row](results).toSet[org.apache.spark.sql.Row]) was false (ContinuousSuite.scala:226)
```

**Comment by HeartSaVioR** [comment]
Looks like I could see the failure locally by on other branch (SPARK-27254). It fails intermittently, but even it succeeds it leaves suspicious error log. I'll see what is happening there.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: Could you put this into another file like `ContinuousTrigger.scala` and the previous `ProcessingTime`?
If possible, please do `git mv` for renaming `ProcessingTime` to `ProcessingTimeTrigger`. Then, update the new file.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: I moved this as @srowen's suggestion (https://github.com/apache/spark/pull/24996#discussion_r299221473) as OneTimeTrigger is there without its own file.

I'm still not sure, but if the intention of deprecation is hiding implementations to end users, actually I'd also like to move ContinuousTrigger to Triggers.scala, as they can be controlled together.

If we change the mind to have file per implementation, Triggers.scala would be better to be renamed as OneTimeTrigger.scala too.

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: Ya. `Moving ContinuousTrigger to Triggers.scala` is also a possible way to be consistent.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: We might need to keep original class as we haven't deprecated it yet, and to allow end users to only create Trigger implementations as `Trigger.xxx`, we may also want to deprecated some more classes as well.

The change may look like below commit:
https://github.com/HeartSaVioR/spark/commit/f8488cfb0a045806e125215b933fc10931b1b431

IMHO this could be considered as another issue as more deprecations are happening. WDYT?

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: Ur, moving looks okay, but the new deprecation of `OneTimeTrigger` is unexpected one for me.
```
@deprecated("use Trigger.Once()", "3.0.0")
// NOTE: In later release, we can change this to `private[sql]` and remove deprecated.
case object OneTimeTrigger extends Trigger
```
Please make another PR for the deprecation of `OneTimeTrigger` if we really need that.

If the PR has multiple themes unexpectedly, we cannot merge it.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: OK I'll make another PR for deprecation of OneTimeTrigger, and change the topic of this PR to "places all Trigger implementations to Triggers.scala". Just pushed the new commit.

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: I think you are moving the opposite direction from the way I recommended. Now, you add a new deprecate in this PR.
```
@deprecated("use Trigger.Continuous(intervalMs)", "3.0.0")
// NOTE: In later release, we can remove this.
```

Why do you add that here? I'm -1 for this PR.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Sorry, @HeartSaVioR . I'm -1 for this PR. This PR is trying to do many things together in the opposite way from the way I recommended. Each item should be considered differently.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: First of all, you recommended opposite direction what I got from @srowen, as I commented earlier, so I just would like to keep this as it is. Let's avoid contributors to do back and forth because of conflict between reviewers.

Second, the new commit reflects your comment like above:

> Moving ContinuousTrigger to Triggers.scala is also a possible way to be consistent.

and I also provided commit and ask for confirmation, and you confirmed.

https://github.com/apache/spark/pull/24996#discussion_r300817205

In confirmation comment I guess you were only concerning changing OneTimeTrigger together, so I excluded it.

Did I miss something?

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: Got it. Please follow PMC's opinion, @HeartSaVioR . Sorry if you feel uncomfortable. I wanted to help and merge this PR, but it seems that it doesn't help you. I'll leave this PR to @srowen fully.

**Comment by dongjoon-hyun** [comment]
Hi, @srowen . Sorry for chiming here. Please review this PR again, @srowen .

**Comment by HeartSaVioR** [comment]
Let me make clear on this (I also talked with @dongjoon-hyun to clear my intention): I didn't decide to keep this as it is because someone is a PMC member and someone is a committer. I kept as it is because it was approved by some committer/PMC member, and applying new recommendation would break it. So I feel we had some miscommunication, and I guess we addressed it.

I'd kindly ask that reviewers would also be open to discuss among them to have consistent direction of recommendations, as contributors are mostly encouraged to apply recommendations if they got some.

Btw, I just rolled back to the status before we got conflict recommendations, so the shape of PR is pretty close to the thing @srowen approved.

**Comment by HeartSaVioR** [comment]
Retest this, please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala
Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: @dongjoon-hyun I don't particularly care where the implementations live, I'm just looking for consistency. It's not terribly consistent to begin with. I think putting the private implementations in `Triggers.scala` is a step in the right direction, given that this file exists at all. I think this change is OK as-is. I don't think we have to move other implementations, though, anything that improves consistency is OK.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala
Comment (tdas) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: Just a flyby comment (apologies if i am missing any prior context or conversation). I do not think it make sense to have both `OnTimeTrigger/ProcessingTimeTrigger` and `Trigger.OneTime/ProcessingTime` be public. The reason I had deprecated `ProcessingTime` constructor is so that we can reduce the number of public classes and move towards having everything through the `Trigger` class using`Trigger.ProcessingTime`. If we are indeed doing that, then we should have `Trigger.ProcessingTime` as the only public method and make `ProcessingTimeTrigger` private/hidden.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala
Comment (tdas) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: Building on the comment I made in a different [thread](https://github.com/apache/spark/pull/24996/files#r299514393), I agree things were not very consistent. Since we are refactoring this, lets make sure things move in the right direction and improve consistency. My opinion on where we should ideally be is 
- Minimize number of public classes that the user needs to know about. That means that ideally there NOT should be multiple public classes and ways for specifying a type of trigger. So there is should be only `Trigger.ProcessingTime` and `ProcessingTimeTrigger` should be hidden.
- Classes that we dont want public visibility should not be visible in Java as well. Making it `public[sql]` still keeps it visible as a top-level in Java API docs. Either move the implementation inside the Trigger class, or move the `ProcessingTimeTrigger` into `sql/execution` package which is explicitly hidden from API docs.

My apologies for missing prior context and conversation. This is just my two cents. I am sure I am probably missing some aspect of the conversation.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: Yeah it fits on my understanding (assumption, actually) of deprecation - let end users only know 
 about `Trigger.xxx` and use these methods instead of instantiating Trigger implementation by theirselves.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: Could we define the boundary of the PR - I guess if we want to address @tdas suggestion, it is going to be more than https://github.com/HeartSaVioR/spark/commit/f8488cfb0a045806e125215b933fc10931b1b431 - just wanted to be clear how much we want to do in this PR.

Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: My suggestion is we follow the comment above by moving `OneTimeTrigger` too, and leave it there. That's more consistent, and consistent with the intent. This leaves the class potentially 'public' in the bytecode, but it was before, and that's true of lots of `private[spark]` classes anyway. While I wouldn't argue with further moving things to `sql/execution`, I think that just putting the implementations in their natural home right now sounds coherent and an improvement, and doesn't expand the change much.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: Sorry I guess I mixed up. My bad, not moving OneTimeTrigger but moving ContinuousTrigger. Let me enumerate necessary changes from what I understand:

* `OneTimeTrigger` is already in `Triggers.scala` but open to `public`, so need to add deprecation and plan to hide it.
  * I guess adding `private[sql]` would work since Triggers.scala is already in `sql.execution.streaming` package. Not sure exact package of `sql.execution` will be hidden.
  * Otherwise we can add replacement as same as what we do for ProcessingTime, but requires bigger changeset.
* Move `ContinuousTrigger` to `Triggers.scala` with `private[sql]` scope, and deprecate origin.

I guess both moving and deprecating make the changeset looking verbose, but I guess even in major release we may not want to remove public classes which haven't been deprecated.

I guess my commit (https://github.com/HeartSaVioR/spark/commit/f8488cfb0a045806e125215b933fc10931b1b431) mentioned above already covered it, so please take a look at the commit. If we are OK to go or would like to continue reviewing under the commit, I'll add the commit to the PR.

Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: Oops I misspoke, I mean "move the implementations to Triggers.scala". 
I'm personally OK with not even deprecating, just moving, as it's a major release and a small detail, but, OK with deprecation too.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: Yeah either is fine for me too. If we would like to have simpler one, skipping deprecation would work. If we would like to have safer one, deprecation would work. I'd like to ask the decision for committers/PMC members, as it seems like related to some policy on the project.

Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: To move this forward, I suggest we just move the class and skip deprecation. A note in the Spark 3.0 migration guide about streaming would be good, as we're removing a deprecated class anyway.

**Comment by HeartSaVioR** [comment]
Just pushed the commit, and adjusted title and content of the PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: Just `object Triggers`? that's the name of the file after all
Do these even need to be moved into their own object? they seem like they're only used in particular classes in this file and aren't abstracting a lot.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: Yeah, renaming is not a big deal. I'll do it once we have decision on below.

Regarding necessity of this object, it abstracts `interval` to provide unified way to convert any kind of intervals what we support, to the internal format of interval. If we want to support other kinds of interval we can add it here, and any new trigger implementations which leverages interval should have benefit on this.

I also hope if we can avoid having same kind of methods among Trigger, companion object of each implementation, and this object. I thought it is just simpler to have this as utility object, but if we want better shape I'll see whether it can be, maybe `trait` would work?

Comment (srowen) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: Yeah OK they're reused across the implementation in this file. Seems OK but I think you can declare it `private` even? I wasn't getting at a `trait` or anything.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala]: > Seems OK but I think you can declare it private even?

Yeah I think that could be. I'll make a change.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
@srowen @tdas Would you mind doing another round of review? Thanks in advance!

**Comment by srowen** [comment]
I think this is OK. We need to add releases notes. I don't think there's a streaming migration guide in the docs, so we can use Docs text in the JIRA. Is this accurate @HeartSaVioR ?

"In Spark 3.0, the deprecated class `org.apache.spark.sql.streaming.ProcessingTime` has been removed. Use `org.apache.spark.sql.streaming.Trigger.ProcessingTime` instead. Likewise, `org.apache.spark.sql.execution.streaming.continuous.ContinuousTrigger` has been removed in favor of `Trigger.Continuous`, and `org.apache.spark.sql.execution.streaming.OneTimeTrigger` has been hidden in favor of `Trigger.Once`."

**Comment by HeartSaVioR** [comment]
> In Spark 3.0, the deprecated class org.apache.spark.sql.streaming.ProcessingTime has been removed. Use org.apache.spark.sql.streaming.Trigger.ProcessingTime instead.

End users are always encouraged to use Trigger.xxx, so this case we need to guide "Use Trigger.ProcessTime"

**Comment by srowen** [comment]
Oh yes, I mean they should use the method `Trigger.ProcessingTime()`; that's not new. (The method's name is unfortunate.)

**Comment by HeartSaVioR** [comment]
Yes I also confused with new class. you're right.

**Comment by srowen** [comment]
Merged to master

**Comment by HeartSaVioR** [comment]
Thanks all for the detailed review and merging!

**Comment by cloud-fan** [comment]
shall we have an item in the migration guide for it?

**Comment by HeartSaVioR** [comment]
Thanks for reminding. We didn't do as migration guide didn't exist for SS. I'll submit a PR quickly.

**Comment by srowen** [comment]
There's no harm in a migration guide note, I think, other than potentially overloading it. This is a case I would have thought release notes cover. What would you write in the migration guide? "use the new class"?

**Comment by HeartSaVioR** [comment]
I guess same content with release note would be OK for migration guide - that's just a matter of preference of references. Suppose end users upgrade to Spark 3.0.0 and find their application fail to compile, which doc they would find for the first time? Migration guide looks to be the centralized one, so maybe preferred over release note.

I have a commit but yet submitted a PR. Please let me know if it makes sense to add it in migration guide.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala:51]: Shall we remove the annotations? it's `private` but the annotations say it's an API.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala:52]: Also, let's don't have `private[sql]` since execution package is already private per SPARK-16964

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Triggers.scala:51]: Ah, sure. Thanks :D.

**Pull request closed without merging**
